### PR TITLE
Add fix_points_in_soma() to remove neurite points located inside the soma

### PR DESCRIPTION
**Pull request by adrien-berchet**

### CODE REVIEW

File: tests/test_sanitize.py [hunk 1]
```diff
@@ -114,3 +118,70 @@ def test_error_annotation_all():
...
+@contextmanager
+def _tmp_file(content, extension):
+    with tempfile.NamedTemporaryFile(suffix='.' + extension, mode='w') as tmp_file:
+        tmp_file.write(content)
+        tmp_file.flush()
+        tmp_file.seek(0)
+
+        yield tmp_file
+
+        tmp_file.close()
```
Comment (adrien-berchet) [inline tests/test_sanitize.py:132]: ups, I forgot to remove this... It was just to test what I should put in the file.

**Review by asanin-epfl — COMMENTED** [review]
please restrict neurom in setup.py to >= 3.0.1

File: neuror/sanitize.py [hunk 2]
```diff
@@ -220,3 +220,40 @@ def annotate_neurolucida_all(morph_paths, nprocesses=1):
...
+
+            last_in_soma = in_soma[-1]
+
+            if last_in_soma >= len(sec_pts) - 1:
+                raise CorruptedMorphology("An entire section is located inside the soma")
+
+            in_pt = sec_pts[last_in_soma]
+            out_pt = sec_pts[last_in_soma + 1]
+            vec = out_pt - in_pt
+            new_pt = [morph.soma.center + vec / np.linalg.norm(vec) * morph.soma.radius]
```
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply neuror/sanitize.py]: Yes but I need a 2D array to concatenate the points, but I changed to a something simpler.

File: neuror/sanitize.py [hunk 3]
```diff
@@ -220,3 +220,40 @@ def annotate_neurolucida_all(morph_paths, nprocesses=1):
...
+                new_sec_pts = root_sec.points[last_in_soma + 1:]
+                last_in_soma += 1
+            else:
+                new_sec_pts = np.concatenate([new_pt, root_sec.points[last_in_soma + 1:]])
+            ptl = PointLevel(
+                new_sec_pts,
+                root_sec.diameters[last_in_soma:],
+                root_sec.perimeters[last_in_soma:],
+            )
+            new_sec = morph.append_root_section(ptl, root_sec.type)
```
[the remark this answers is not in the document]
Comment (adrien-berchet) [reply neuror/sanitize.py]: Ah yes, I thought the points were immutable. Thanks!

**Review by asanin-epfl — APPROVED** [review]
perfetto!

File: neuror/sanitize.py [hunk 4]
```diff
@@ -220,3 +220,34 @@ def annotate_neurolucida_all(morph_paths, nprocesses=1):
             morph_path = str(morph_path)
             annotations[morph_path], summaries[morph_path], markers[morph_path] = result
     return annotations, summaries, markers
+
+
+def fix_points_in_soma(morph):
+    """Ensure section points are not inside the soma."""
```
Comment (arnaudon) [inline neuror/sanitize.py:226]: maybe describe a bit what it does if there are some points in the soma?

**Review by arnaudon — APPROVED** [review]

File: neuror/sanitize.py [hunk 5]
```diff
@@ -223,7 +223,15 @@ def annotate_neurolucida_all(morph_paths, nprocesses=1):
...
 
 def fix_points_in_soma(morph):
-    """Ensure section points are not inside the soma."""
+    """Ensure section points are not inside the soma.
+
+    Method:
+        - for each root section, we check which points are inside the soma.
+        - if all points of a root section are inside the soma, an exception is raised because it
+        means that a bifurcation is located inside the soma, which is hard to automatically fix.
+        - if there is at least 1 point inside the soma, a new point is defined to replace of them.
```
Comment (arnaudon) [inline neuror/sanitize.py:232]: to replace them

Comment (adrien-berchet) [reply neuror/sanitize.py:232]: ups :-)

**Comment by asanin-epfl** [comment]
ok, looks good

**Comment by asanin-epfl** [comment]
if you plan to release then use 1.5.0 version

**Comment by adrien-berchet** [comment]
ok :+1:

**Pull request merged**